Postgresql: Support for TimescaleDB

tbutts opened the pull request:
Hi there, this PR is my attempt to add support in testfixtures for [TimescaleDB](https://www.timescale.com/), a Postgres extension that supports time-series operations & workflows. I really like the idea of testing with a live db data especially when there's complex db functions involved, so testfixtures was a great find. Unfortunately, this library breaks when attempting to restage the timescaledb tables with the error `ERROR: operation not supported on chunk tables`.

TimescaleDB works transparently by manipulating the database, automatically generating partition tables, hooking into the query planner, and other tricks.

```
psql~> \dn
          List of schemas
          Name           |  Owner   
-------------------------+----------
 _timescaledb_cache      | postgres
 _timescaledb_catalog    | postgres
 _timescaledb_config     | postgres
 _timescaledb_internal   | postgres
 public                  | postgres
```
My approach to fix the issue I saw, was to ignore those schemas and let the extension do the book-keeping. Specifically, I added an exclusion for all pg namespaces that are prefixed by underscores (_), which are considered private, off-limits by the extension. This may be kludgy and break existing user's tests if they happen to prefix similar ways, though I've never personally seen that naming strategy used by anyone but applications that mess with postgres internals.

I would be willing to investigate an alternative approach. Such as searching for partition tables that do not support DELETE/TRUNCATE, or adding an option field to the `testfixtures.Helper` to specify a blacklist of table namespaces. Let me know what you think!

---

Another user wondered about Timescale before, #27 . Around that time, I started using this patch with my own projects, and I have seen no unexpected breaks. Testing was done with PG v10 & v11, and the extension for TimescaleDB v0.9 (beta) through v1.4 (current).

tbutts committed:
Postgres: Ignore 'internal' namespaces during Load

This change fixed a number of odd bugs that can occur when using
testfixtures with `timescaledb` - a Postgres extension for time-series
data. Timescale partitions tables automatically to enable
high-throughput and query optimization. Timescale tables are placed into
a few namespaces which all start with an underscore (_), which are
considered private, off-limits by the extension. This commit causes
testfixtures to ignore tables in namespaces that start with an
underscore.

This change may break other applications that generate tables in
similarly named namespaces.

andreynering commented:
Hi @tbutts, thanks for opening this PR!

Do you know if TimescaleDB can have more schemas, or only those 4? If we know the exact schema names, I'd vote to exclude those 4 schemas from the query instead if every schema which starts with `_`. Alternatively, we could exclude those which starts with `_timescaledb`. What do you think?

Also, can you change [this title](https://github.com/go-testfixtures/testfixtures/blob/master/README.md#postgresql) from "PostgreSQL" to "PostgreSQL / TimescaleDB"?

tbutts commented:
Hey thanks for taking the time to look, I appreciate it!

I did some research on the schema/namespace names, especially since I wasn't sure if the prefixed underscore(_) names were a convention or a Postgres requirement. Looking at the docs[[0]][[1]] and on stackoverflow, there's nothing reserved about Timescale's schemas. Since there's no easy way to predict this for other extensions, we should narrow it to support timescale specifically. While I agree there's safety & simplicity in spelling out the four schemas, from what I see in timescale's primer SQL scripts they did add a schema last year[[2]] (Aug 3, 2018). They could change more, but if you're fine with the risk of some minor maintenance, then I'll do whichever method you prefer.

I'll update the title, too. Afterwards, do you want me to rebase my changes or leave the commits in?

[0]: https://www.postgresql.org/docs/12/sql-syntax-lexical.html
[1]: https://www.postgresql.org/docs/12/ddl-schemas.html
[2]: https://github.com/timescale/timescaledb/blame/ddb8f46b5f0137bf6d65b3b743ff62b7416a2860/sql/pre_install/schemas.sql

andreynering commented:
I think excluding `_timescaledb%` is the best balance, then. More future-proof in case they add more schemas, but also won't affect user schemas with underscores (or potentially other extensions).

And it's fine to just add more commits. 🙂

tbutts committed:
Postgres: Specifically ignore _timescaledb on Load; Update readme

Per feedback on PR #53 , testfixtures will support timescaledb by
explicitly ignoring its schemas and letting the extension manage itself.
Any schemas prefixed with `_timescaledb` will be left alone.
Due to the hooks into PG that TimescaleDB uses, it is invalid to try to
mess with the automatic partition tables, caching, and other data the
extension keeps.

Also, the readme now calls out support for TimescaleDB. This helps
others know that this library will support their architecture.

tbutts commented:
All ready to merge when you want. Thanks again, for all your hard work!

andreynering commented:
@tbutts Thanks a lot for your contribution!

I'm gonna merge and release this.
